numba/core/ssa.py

```diff
@@ -0,0 +1,341 @@
+        varnames = [k.name for k in used_vars]
+        phivar = states['varname']
+        if phivar in varnames:
+            return self._find_def(states, stmt)
+
+    def _find_def(self, states, stmt):
+        _logger.debug("find_def var=%r stmt=%s", states['varname'], stmt)
+        selected_def = None
+        label = states['label']
+        local_defs = states['defmap'][label]
+        local_phis = states['phimap'][label]
+        block = states['block']
+
+        cur_pos = self._stmt_index(stmt, block)
+        for defstmt in reversed(local_defs):
+            # Phi nodes have no index
+            def_pos = self._stmt_index(defstmt, block, stop=cur_pos)
+            if def_pos < cur_pos:
+                selected_def = defstmt
+                break
+            # Maybe it's a PHI
+            elif defstmt in local_phis:
+                selected_def = local_phis[-1]
+                break
+
+        if selected_def is None:
+            selected_def = self._find_def_from_top(
+                states, label, loc=stmt.loc,
+            )
+        return selected_def
+
+    def _find_def_from_top(self, states, label, loc):
+        _logger.debug("find_def_from_top label %r", label)
+        cfg = states['cfg']
+        defmap = states['defmap']
+        phimap = states['phimap']
+        domfronts = states['df+']
+        for deflabel, defstmt in defmap.items():
+            df = domfronts[deflabel]
+            if label in df:
+                scope = states['scope']
+                loc = states['block'].loc
+                # fresh variable
+                freshvar = scope.redefine(states['varname'],
+                                          loc=loc)
+                # insert phi
+                phinode = ir.Assign(
+                    target=freshvar,
+                    value=ir.Expr.phi(loc=loc),
+                    loc=loc,
+                )
+                _logger.debug("insert phi node %s at %s", phinode, label)
+                defmap[label].insert(0, phinode)
+                phimap[label].append(phinode)
+                for pred, _ in cfg.predecessors(label):
+                    incoming_def = self._find_def_from_bottom(
+                        states, pred, loc=loc,
+                    )
+                    _logger.debug("incoming_def %s", incoming_def)
+                    phinode.value.incoming_values.append(incoming_def.target)
+                    phinode.value.incoming_blocks.append(pred)
+                return phinode
+        else:
+            idom = cfg.immediate_dominators()[label]
+            if idom == label:
+                # We are searched to the top of the idom tree.
+                # Since we still cannot find a definition,
+                # we will warn and let
+                warnings.warn(
+                    errors.NumbaWarning(
+                        f"Detected uninitialized variable {states['varname']}",
+                        loc=loc),
+                )
```

Comment:
As discussed in the PR feedback, this needs to be behind a config option as it's potentially confusing.